Debuggable IO [DIO]

kaishh opened the pull request:
ZIO Tracing's True Power, as its name suggests, is to reign supreme over the entire world!

![big_1459812265_image](https://user-images.githubusercontent.com/450507/57573359-6ee73d80-741e-11e9-9b36-224aa93f39b8.jpg)

iravid commented:
@kaishh congrats! impressive work! 

Is https://github.com/lihaoyi/sourcecode not used to avoid adding macros to the project?

kaishh commented:
@iravid that, and compile-time solutions cannot provide support for 3rd-party libraries, like doobie, http4s and existing code written with tagless final so we decided to go with the other way

iravid commented:
Cool, did not consider that! 👍

wi101 commented on core/shared/src/main/scala/scalaz/zio/ZTrace.scala:
```diff
@@ -0,0 +1,36 @@
+        s"\nFiber:$fiberId was supposed to continue to:" ::
+          this.stackTrace.reverse.map(loc => s"  a future continuation at " + loc.prettyPrint)
+      else
+        s"\nFiber:$fiberId was supposed to continue to: <empty trace>" ::Nil
```
I think, format code `fmt` is missing

almendar commented on core/shared/src/main/scala/scalaz/zio/internal/tracing/ZIOFn.scala:
```diff
@@ -0,0 +1,36 @@
+ * Allows access to the underlying lambda so that stacktraces can
+ * point to supplied user's code, instead of ZIO internals.
+ *
+ * NOTE: it being an `abstract class`, not trait is important as typecasing
```
```suggestion
 * NOTE: it being an `abstract class`, not trait is important as typecasting
```

kaishh replied:
`type casing` / `type case` is actually a valid term for matching on types, that was not a typo!

almendar replied:
TIL 😄 
Sorry I didn't know and spellchecker in IDEA underlined it.

ghostdogpr commented:
@kaishh 👏this looks brilliant! Can't wait to try it on a real project.

pshirshov commented on core/shared/src/main/scala/scalaz/zio/internal/SingleThreadedRingBuffer.scala:
```diff
@@ -0,0 +1,53 @@
+  }
+
+  @inline private[this] def decrement(): Unit =
+    if (idx != 0) {
```
Issue here.

thiloplanz commented:
I have a question: Could the mechanism that is now in place to trace execution also support collecting profiling information (i.e. how long each effect took to complete)? If so, would this be effectively free or add significant overhead?

artempyanykh commented on core/shared/src/main/scala/scalaz/zio/internal/SingleThreadedRingBuffer.scala:
```diff
@@ -0,0 +1,53 @@
+package scalaz.zio.internal
+
+private[zio] final class SingleThreadedRingBuffer[A <: AnyRef](capacity: Int) {
```
A couple questions regarding semantics of this class:
1. Is it intended that elements get overwritten on buffer wrap around?
2. SingleThreaded = NotThreadSafe?
3. `pop` returning Unit is unusual. Could you elaborate?

All in all, it would be nice to add a scaladoc to this class documenting its semantics. Could you?

pshirshov replied:
1. Yes, it's a ring buffer
2. Yes and it's fine for the purpose. Other fiber structures are the same.
3. It's not 'pop', it's 'remove head'.

kaishh replied:
2. We have doubts about thread safety here, however, both `Stack` & `StackBool` are currently implemented the same way, so I guess we'll find out sooner or later...

artempyanykh commented on core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala:
```diff
@@ -49,6 +54,74 @@ private[zio] final class FiberContext[E, A](
   private[this] val locked          = Stack[Executor]()
   private[this] val supervised      = Stack[Set[FiberContext[_, _]]]()
 
+  private[this] val tracingStatus = if (tracingEnabled) StackBool(initialTracingStatus) else null
```
Could you add a blank line in-between those? Otherwise a bit hard to read.

kaishh replied:
done!

artempyanykh commented on stacktracer/jvm/src/main/scala/scalaz/zio/internal/stacktracer/impl/AkkaTracer.scala:
```diff
@@ -0,0 +1,32 @@
+
+import scala.util.matching.Regex
+
+final class AkkaTracer extends Tracer {
```
Is it called `AkkaTracer` just to give credit (since it uses AkkaLineNumber under the hood)? My first reaction to this name was 'oh, does zio depend on akka now'. Maybe something like `AkkaInspiredTracer` would be a clearer name?

kaishh replied:
I think there's little point in overt distancing, overly-precise terms, underscoring of differences or anything that could provoke more drama. This code isn't inspired, but straight up stolen 😉 and we're probably going to benefit from updates to the upstream version in the future, so I think it's ok to call it what it is.

artempyanykh replied:
@kaishh I didn't have any of that in mind. I just shared that it was not clear to me that `AkkaTracer` is a standalone component and not an unexpected dependency on akka or akka-specific component. Anyway, I don't feel strongly about the names, but hoping you'll provide some context in the class doc.

kaishh replied:
Fair enough! I've renamed it to `AkkaLineNumbersTracer` and qualified that it merely uses akka's bytecode parser, without incuring an akka dependency.

artempyanykh commented on stacktracer/jvm/src/main/scala/scalaz/zio/internal/stacktracer/impl/AkkaLineNumbers.scala:
```diff
@@ -0,0 +1,374 @@
+ * this particular lambda (and there might be multiple defined within a single
+ * class).
+ **/
+object AkkaLineNumbers {
```
Could you please provide tracer benchmark numbers in a PR comment? It's not clear right away what are the benefits of using this code instead of ASM based parser (outside of maintaining a zero dep status).

pshirshov replied:
Asm is a bih faster but this parser has no external dependencies. And this is the primary reason to choose this.

Also keep in mind that once the cache is filled this code will never be invoked.

dnaumenko replied:
@pshirshov The cache is unbounded and could be a potential problem on large codebases. Do you have plans to optimize it? One of optimization from top of my head is to use String.intern() method for file, clazz maybe

pshirshov replied:
Check the whole patch, interning is there already.

dnaumenko replied:
Didn't notice it, my bad. Cool!

kaishh replied:
@artempyanykh Benchmark numbers on my machine:
```
[info] Running (fork) org.openjdk.jmh.Main scalaz.zio.stacktracer.TracersBenchmark -i 10 -wi 10 -f1 -t8
[info] # JMH version: 1.21
[info] # VM version: JDK 1.8.0_181, Java HotSpot(TM) 64-Bit Server VM, 25.181-b13
[info] # VM invoker: /Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/bin/java
[info] # VM options: <none>
[info] # Warmup: 10 iterations, 1 s each
[info] # Measurement: 10 iterations, 1 s each
[info] # Timeout: 10 min per iteration
[info] # Threads: 8 threads, will synchronize iterations
[info] # Benchmark mode: Average time, time/op
[info] # Benchmark: scalaz.zio.stacktracer.TracersBenchmark.akkaLineNumbers1000Lambdas
[info] # Run progress: 0.00% complete, ETA 00:01:00
[info] # Fork: 1 of 1
[info] # Warmup Iteration   1: 688.078 ?(99.9%) 333.396 ms/op
[info] # Warmup Iteration   2: 539.822 ?(99.9%) 10.256 ms/op
[info] # Warmup Iteration   3: 526.389 ?(99.9%) 10.554 ms/op
[info] # Warmup Iteration   4: 477.833 ?(99.9%) 17.939 ms/op
[info] # Warmup Iteration   5: 481.406 ?(99.9%) 16.112 ms/op
[info] # Warmup Iteration   6: 480.685 ?(99.9%) 19.045 ms/op
[info] # Warmup Iteration   7: 493.985 ?(99.9%) 22.763 ms/op
[info] # Warmup Iteration   8: 490.212 ?(99.9%) 15.283 ms/op
[info] # Warmup Iteration   9: 502.920 ?(99.9%) 29.178 ms/op
[info] # Warmup Iteration  10: 497.979 ?(99.9%) 25.670 ms/op
[info] Iteration   1: 506.635 ?(99.9%) 26.874 ms/op
[info] Iteration   2: 511.377 ?(99.9%) 34.289 ms/op
[info] Iteration   3: 519.323 ?(99.9%) 42.639 ms/op
[info] Iteration   4: 499.538 ?(99.9%) 29.123 ms/op
[info] Iteration   5: 517.134 ?(99.9%) 22.145 ms/op
[info] Iteration   6: 498.225 ?(99.9%) 17.942 ms/op
[info] Iteration   7: 526.698 ?(99.9%) 15.620 ms/op
[info] Iteration   8: 502.834 ?(99.9%) 18.593 ms/op
[info] Iteration   9: 522.352 ?(99.9%) 23.818 ms/op
[info] Iteration  10: 490.843 ?(99.9%) 16.605 ms/op
[info] Result "scalaz.zio.stacktracer.TracersBenchmark.akkaLineNumbers1000Lambdas":
[info]   509.496 ?(99.9%) 17.791 ms/op [Average]
[info]   (min, avg, max) = (490.843, 509.496, 526.698), stdev = 11.768
[info]   CI (99.9%): [491.705, 527.287] (assumes normal distribution)
[info] # JMH version: 1.21
[info] # VM version: JDK 1.8.0_181, Java HotSpot(TM) 64-Bit Server VM, 25.181-b13
[info] # VM invoker: /Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/bin/java
[info] # VM options: <none>
[info] # Warmup: 10 iterations, 1 s each
[info] # Measurement: 10 iterations, 1 s each
[info] # Timeout: 10 min per iteration
[info] # Threads: 8 threads, will synchronize iterations
[info] # Benchmark mode: Average time, time/op
[info] # Benchmark: scalaz.zio.stacktracer.TracersBenchmark.akkaTracer1000Lambdas
[info] # Run progress: 33.33% complete, ETA 00:01:11
[info] # Fork: 1 of 1
[info] # Warmup Iteration   1: 810.213 ?(99.9%) 344.045 ms/op
[info] # Warmup Iteration   2: 610.676 ?(99.9%) 32.374 ms/op
[info] # Warmup Iteration   3: 557.663 ?(99.9%) 30.693 ms/op
[info] # Warmup Iteration   4: 583.463 ?(99.9%) 24.476 ms/op
[info] # Warmup Iteration   5: 526.098 ?(99.9%) 19.209 ms/op
[info] # Warmup Iteration   6: 553.463 ?(99.9%) 12.259 ms/op
[info] # Warmup Iteration   7: 539.786 ?(99.9%) 22.442 ms/op
[info] # Warmup Iteration   8: 553.602 ?(99.9%) 21.540 ms/op
[info] # Warmup Iteration   9: 535.725 ?(99.9%) 26.722 ms/op
[info] # Warmup Iteration  10: 556.537 ?(99.9%) 18.868 ms/op
[info] Iteration   1: 536.602 ?(99.9%) 12.564 ms/op
[info] Iteration   2: 558.272 ?(99.9%) 16.455 ms/op
[info] Iteration   3: 525.707 ?(99.9%) 19.731 ms/op
[info] Iteration   4: 559.580 ?(99.9%) 29.885 ms/op
[info] Iteration   5: 532.769 ?(99.9%) 20.721 ms/op
[info] Iteration   6: 556.363 ?(99.9%) 8.802 ms/op
[info] Iteration   7: 565.317 ?(99.9%) 15.784 ms/op
[info] Iteration   8: 576.543 ?(99.9%) 9.204 ms/op
[info] Iteration   9: 548.832 ?(99.9%) 24.991 ms/op
[info] Iteration  10: 578.714 ?(99.9%) 13.820 ms/op
[info] Result "scalaz.zio.stacktracer.TracersBenchmark.akkaTracer1000Lambdas":
[info]   553.870 ?(99.9%) 27.057 ms/op [Average]
[info]   (min, avg, max) = (525.707, 553.870, 578.714), stdev = 17.897
[info]   CI (99.9%): [526.813, 580.927] (assumes normal distribution)
[info] # JMH version: 1.21
[info] # VM version: JDK 1.8.0_181, Java HotSpot(TM) 64-Bit Server VM, 25.181-b13
[info] # VM invoker: /Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/bin/java
[info] # VM options: <none>
[info] # Warmup: 10 iterations, 1 s each
[info] # Measurement: 10 iterations, 1 s each
[info] # Timeout: 10 min per iteration
[info] # Threads: 8 threads, will synchronize iterations
[info] # Benchmark mode: Average time, time/op
[info] # Benchmark: scalaz.zio.stacktracer.TracersBenchmark.asmTracer1000Lambdas
[info] # Run progress: 66.67% complete, ETA 00:00:34
[info] # Fork: 1 of 1
[info] # Warmup Iteration   1: 392.895 ?(99.9%) 175.151 ms/op
[info] # Warmup Iteration   2: 335.377 ?(99.9%) 53.985 ms/op
[info] # Warmup Iteration   3: 673.830 ?(99.9%) 25.490 ms/op
[info] # Warmup Iteration   4: 877.822 ?(99.9%) 194.560 ms/op
[info] # Warmup Iteration   5: 329.584 ?(99.9%) 16.764 ms/op
[info] # Warmup Iteration   6: 229.133 ?(99.9%) 10.911 ms/op
[info] # Warmup Iteration   7: 249.190 ?(99.9%) 13.161 ms/op
[info] # Warmup Iteration   8: 224.898 ?(99.9%) 55.644 ms/op
[info] # Warmup Iteration   9: 279.752 ?(99.9%) 15.074 ms/op
[info] # Warmup Iteration  10: 207.423 ?(99.9%) 5.639 ms/op
[info] Iteration   1: 240.831 ?(99.9%) 12.544 ms/op
[info] Iteration   2: 228.104 ?(99.9%) 13.935 ms/op
[info] Iteration   3: 181.641 ?(99.9%) 4.350 ms/op
[info] Iteration   4: 194.082 ?(99.9%) 14.451 ms/op
[info] Iteration   5: 262.669 ?(99.9%) 13.056 ms/op
[info] Iteration   6: 241.883 ?(99.9%) 9.852 ms/op
[info] Iteration   7: 210.150 ?(99.9%) 3.828 ms/op
[info] Iteration   8: 359.880 ?(99.9%) 33.290 ms/op
[info] Iteration   9: 209.194 ?(99.9%) 5.922 ms/op
[info] Iteration  10: 260.227 ?(99.9%) 9.506 ms/op
[info] Result "scalaz.zio.stacktracer.TracersBenchmark.asmTracer1000Lambdas":
[info]   238.866 ?(99.9%) 76.064 ms/op [Average]
[info]   (min, avg, max) = (181.641, 238.866, 359.880), stdev = 50.312
[info]   CI (99.9%): [162.802, 314.930] (assumes normal distribution)
[info] # Run complete. Total time: 00:01:38
[info] REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
[info] why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
[info] experiments, perform baseline and negative tests that provide experimental control, make sure
[info] the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
[info] Do not assume the numbers tell you what you want them to tell.
[info] Benchmark                                    Mode  Cnt    Score    Error  Units
[info] TracersBenchmark.akkaLineNumbers1000Lambdas  avgt   10  509.496 ? 17.791  ms/op
[info] TracersBenchmark.akkaTracer1000Lambdas       avgt   10  553.870 ? 27.057  ms/op
[info] TracersBenchmark.asmTracer1000Lambdas        avgt   10  238.866 ? 76.064  ms/op
[success] Total time: 112 s, completed 22-May-2019 03:01:35
```
Tracer from akka-typed is about ~2x slower, but has no dependencies.

artempyanykh commented on stacktracer/shared/src/main/scala/scalaz/zio/internal/stacktracer/Tracer.scala:
```diff
@@ -0,0 +1,30 @@
+}
+
+object Tracer {
+  def globallyCached(tracer: Tracer): Tracer = cached(globalMutableSharedTracerCache)(tracer)
```
I'm not sure about the utility of this method. It's used only once in `PlatformLive` initialization, and it's not clear when you'd want several tracers that share a single global cache.

kaishh replied:
I'd say you _always_ want multiple tracers to share a global cache. e.g. if you're running tests in parallel with multiple RTS's if you're using a global cache you can go through the cache warm up for the common code just once, instead of all over again for every test. Since the extracted information is final and determined after class loading, the results from tracing cannot change - as such it nearly never really makes sense to NOT use a global cache

artempyanykh replied:
All righty! I guess the _multiple RTSes_ use-case wasn't clear to me.

artempyanykh commented on stacktracer/shared/src/main/scala/scalaz/zio/internal/stacktracer/Tracer.scala:
```diff
@@ -0,0 +1,30 @@
+    }
+
+  final type TracerCache = ConcurrentHashMap[Class[_], ZTraceElement]
+  lazy val globalMutableSharedTracerCache = new TracerCache(10000)
```
Can we make this private?

kaishh replied:
done!

pshirshov commented:
@thiloplanz I think we may collect metrics here and it should be cheap enough. Though these metrics may be a bit cryptic and not so easy to interpret because of the first line issue and the nature of for comprehension. Still may be worthy to try.

guizmaii commented on benchmarks/src/main/scala/scalaz/zio/stacktracer/impls/AsmTracer.scala:
```diff
@@ -0,0 +1,83 @@
+    val clazz = lambda.getClass
+    val classLoader = clazz.getClassLoader
+
+    Try(clazz.getDeclaredMethod("writeReplace")) match {
```
Why do you use a `Try` and not a `try catch` which seems more performant (less allocations) ?

ghostdogpr commented on core/shared/src/main/scala/scalaz/zio/Exit.scala:
```diff
@@ -282,23 +285,43 @@ object Exit extends Serializable {
 
+    /**
+     * Remove all `Fail` and `Interrupt` nodes from this `Cause`,
+     * return only [[Die]] cause/finalizer defects.
```
Should be `` `Die` `` otherwise Scaladoc complains.

kaishh replied:
done!

ghostdogpr commented:
It would be nice to be able to filter out some parts of the stacks. I tried it on my work project and I see a LOT of noise (> 75% of it) generated by internal zio combinators, cats and cats effect interop, http4s, etc.

regiskuckaertz commented on core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala:
```diff
@@ -145,36 +234,53 @@ private[zio] final class FiberContext[E, A](
                       val io2 = nested.asInstanceOf[ZIO.Succeed[Any]]
 
-                      curIo = io.k(io2.value)
+                      if (traceExec && inTracingRegion) addTrace(k)
```
That may be a silly question but what is the cost of all these checks when everything is disabled? Perhaps it is worth turning adding a nanopass AST phase to the compiler that is the identity function in `FiberContext` and one that calls `addTrace` in `FiberTracing` or something like that.

kaishh replied:
It's written like that so that the JIT can compile it away completely when it sees that it will never be true because a `final` variable is false / was false for a long time. Benchmarks actually show an *increase* in performance with disabled compared to master because I turned inlining back on in sbt. I think anything involving closures might likely be slower than a conditional.

pshirshov replied:
`if` should be lot faster than a stackframe creation. Also we checked and there is no noticeable perf impact when tracing is off.

regiskuckaertz replied:
Thanks both, that's very insightful 👍

kaishh commented:
I'll need to do some firefighting this week, so might probably be able to find some time to finish cleaning and refactoring (and addressing past feedback by John) before next week, but likely not before.

pshirshov commented:
> It would be nice to be able to filter out some parts of the stacks. I tried it on my work project and I see a LOT of noise (> 75% of it) generated by internal zio combinators, cats and cats effect interop, http4s, etc.

Once this P/R is merged I will provide another one with some opt-in deduplication and cleanup heuristics

kaishh commented on benchmarks/src/main/scala/scalaz/zio/stacktracer/impls/AsmTracer.scala:
```diff
@@ -0,0 +1,83 @@
+      if (name == methodName && (methodSignature == null || methodSignature == descriptor)) {
+        new MethodVisitor(apiVersion) {
+          override def visitLineNumber(line: Int, start: Label): Unit =
+            //println((line, start))
```
done!

kaishh commented:
@jdegoes @artempyanykh Guess I'm ready for another review

jdegoes commented on core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala:
```diff
@@ -110,6 +204,26 @@ private[zio] final class FiberContext[E, A](platform: Platform, startEnv: AnyRef
+
+    // Store the trace of the immediate future flatMap during evaluation
+    // of a 1-hop left bind, to show a stack trace closer to the point of failure
+    var fastPathFlatMapContinuationTrace: ZTraceElement = null
```
This will be allocated on the heap in an `ObjectRef`.

jdegoes replied:
Which argues for pulling this logic out into the aforementioned utility class.

kaishh replied:
Are you sure? I'm assigning to a val before passing. Also null can't be allocated on the heap?

kaishh replied:
Ok, I see that

kaishh replied:
inlining forces that to disappear, but gives, consistently, just slightly lower performance:
`@noinline` with ObjectRef:
```
[info] Result "scalaz.zio.IOLeftBindBenchmark.scalazLeftBindBenchmark":
[info]   4930.188 ±(99.9%) 77.862 ops/s [Average]
[info]   (min, avg, max) = (4817.133, 4930.188, 4987.029), stdev = 51.501
[info]   CI (99.9%): [4852.325, 5008.050] (assumes normal distribution)
```
`@inline` without ObjectRef:
```
[info] Result "scalaz.zio.IOLeftBindBenchmark.scalazLeftBindBenchmark":
[info]   4893.806 ±(99.9%) 74.707 ops/s [Average]
[info]   (min, avg, max) = (4821.052, 4893.806, 4989.764), stdev = 49.414
[info]   CI (99.9%): [4819.099, 4968.512] (assumes normal distribution)
```
'worker-wrapper' with `@noinline`  on part of fastPathTrace and inlined assignment without ObjectRef
```
[info] Result "scalaz.zio.IOLeftBindBenchmark.scalazLeftBindBenchmark":
[info]   4887.192 ±(99.9%) 105.314 ops/s [Average]
[info]   (min, avg, max) = (4777.507, 4887.192, 5003.847), stdev = 69.658
[info]   CI (99.9%): [4781.878, 4992.505] (assumes normal distribution)
```

Seems like the gain from `@noinline` hot-spot is consistently higher here than the overhead of ObjectRef

jdegoes commented on core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala:
```diff
@@ -144,31 +259,55 @@ private[zio] final class FiberContext[E, A](platform: Platform, startEnv: AnyRef
-
-                      stack.push(io.k)
+                      if (traceStack && inTracingRegion) addStackTrace(k)
+                      stack.push(k)
```
After seeing this pattern repeated multiple times, I'm more convinced we need a utility method that can call `addStackTrace(k)` and `stack.push(k)` at the same time.

kaishh replied:
:+1:

jdegoes commented on core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala:
```diff
@@ -179,55 +318,73 @@ private[zio] final class FiberContext[E, A](platform: Platform, startEnv: AnyRef
+                  val effect = io.effect
 
-                  curIo = nextInstr(io.effect())
+                  if (traceEffects && inTracingRegion) addTrace(effect)
```
All the testing for `inTracingRegion` can be accelerated (right now it goes through heap access every time).

kaishh replied:
Will do. Although, I have to recheck, but I'm fairly sure it gets JIT'ed away on current benchmarks due to tracingStatus never changing, will need a benchmark that toggles tracing

kaishh replied:
After testing: Putting the StackBool on stack hurts performance in the common case when tracing status is not changed. Using a `var currentStatus: Boolean` instead of the heap also hurts perf (4800 vs. 4900 ops). Seems like using class fields is more jit friendly in this case.

kaishh replied:
Using a `var` means we have to create a `BooleanRef` to be able to change it in `bracket_` which hurts performance and JIT-ability... Adding more primops to toggle status might alleviate that, but I guess it's too much since we're probably getting the best performance for the commont case of non-changing tracing status currently anyway.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZIO.scala:
```diff
@@ -866,7 +891,7 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
    * }}}
    */
-  final def tap[R1 <: R, E1 >: E](f: A => ZIO[R1, E1, _]): ZIO[R1, E1, A] = self.flatMap(a => f(a).const(a))
+  final def tap[R1 <: R, E1 >: E](f: A => ZIO[R1, E1, _]): ZIO[R1, E1, A] = self.flatMap(new ZIO.TapFn(f))
```
All of this manual wrapping will be hard to maintain and easy to get wrong (plus, of course, it adds overhead that cannot be escaped).

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZIO.scala:
```diff
@@ -2060,4 +2196,11 @@ object ZIO extends ZIO_R_Any {
   private[zio] final class FiberRefModify[A, B](val fiberRef: FiberRef[A], val f: A => (B, A)) extends UIO[B] {
     override def tag = Tags.FiberRefModify
   }
+  private[zio] final class Trace[R] extends ZIO[R, Nothing, ZTrace] {
```
`object`?

kaishh replied:
:+1:

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZIO.scala:
```diff
@@ -2060,4 +2196,11 @@ object ZIO extends ZIO_R_Any {
+    override def tag = Tags.Trace
+  }
+
+  private[zio] final class TracingStatus[R, E, A](val zio: ZIO[R, E, A], val flag: Boolean) extends ZIO[R, E, A] {
```
For completeness, would be good to have a way to check tracing status, like `checkInterrupt`.

kaishh replied:
:+1:

kaishh commented on core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala:
```diff
@@ -41,13 +49,89 @@ private[zio] final class FiberContext[E, A](platform: Platform, startEnv: AnyRef
+    if (traceStack) SingleThreadedRingBuffer[ZTraceElement](platform.tracing.tracingConfig.stackTraceLength)
+    else null
+
+  private[this] val tracer = platform.tracing.tracer
```
:+1:

kaishh replied:
Moved to https://github.com/scalaz/scalaz-zio/issues/915

kaishh commented on core/shared/src/main/scala/scalaz/zio/internal/SingleThreadedRingBuffer.scala:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  @inline private[this] final def decrement(): Unit = {
+//    assert(size > 0 && current >= 0 && current < capacity)
```
yes, will remove

kaishh commented on core/shared/src/main/scala/scalaz/zio/internal/tracing/ZIOFn.scala:
```diff
@@ -0,0 +1,35 @@
+ * on class appears to be about 10 times faster in [[scalaz.zio.internal.FiberContext.unwrap]]
+ * hot-spot.
+ * */
+private[zio] abstract class ZIOFn extends Serializable {
```
On heap allocations, only the second approach will truly result in less allocations, as the first approach with .traced would have to allocate a new ZIO op. On the first approach, not sure how to replace nested wrappers with it (e.g. in bracket* family). It's definitely worth a try to replace ZIOFn scheme, although I think to postpone it from the current PR as its non-trivial, and a fix for it would not break bincompat.

kaishh replied:
I will definitely try to implement second scheme today, if not succesful - postpone

kaishh replied:
I've replaced some `ZIOFn` uses with a nullable field in `FlatMap` in [another branch](https://github.com/7mind/scalaz-zio/compare/feature/zio-stacktraces...feature/remove-ziofn) - however, I'm not seeing any performance improvement from less allocations in the benchmarks

kaishh replied:
As such, punting this for now

kaishh replied:
Continued here - https://github.com/scalaz/scalaz-zio/issues/916

kaishh commented on interop-java/jvm/src/test/scala/scalaz/zio/interop/javaconcurrentSpec.scala:
```diff
@@ -54,13 +54,13 @@ class javaconcurrentSpec(implicit ee: ExecutionEnv) extends TestRuntime {
     val ex                     = new Exception("no future for you!")
     def noFuture: Future[Unit] = throw ex
-    unsafeRun(Task.fromFutureJava(noFuture _)) must (throwA(FiberFailure(Die(ex))))
+    unsafeRunSync(Task.fromFutureJava(noFuture _)) must_=== Exit.Failure(die(ex))
```
Unfortunately, `throwA` does not use equals, it compares the output of `.getMessage` which is unavoidably different when traces and fiberIds are involved!

kaishh commented on core/shared/src/main/scala/scalaz/zio/internal/SingleThreadedRingBuffer.scala:
```diff
@@ -0,0 +1,61 @@
+package scalaz.zio.internal
```
:+1:

kaishh commented on core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala:
```diff
@@ -110,6 +204,26 @@ private[zio] final class FiberContext[E, A](platform: Platform, startEnv: AnyRef
+    // Put tracing configuration on the stack:
+//    val traceExec    = this.traceExec
+//    val traceStack   = this.traceStack
+//    val traceEffects = this.traceEffects
```
Yes. Placing them on stack turns out to hurt JIT instead of helping it, I'll remove the comments.

kaishh commented on core/shared/src/main/scala/scalaz/zio/Exit.scala:
```diff
@@ -412,17 +471,24 @@ object Exit extends Serializable {
+    final case class Traced[E](cause: Cause[E], trace: ZTrace) extends Cause[E] {
+      // Traced is excluded completely from equals & hashCode
+      override final def hashCode(): Int = cause.hashCode()
+      override final def equals(obj: Any): Boolean = obj match {
```
:+1:

kaishh commented on core/js/src/main/scala/scalaz/zio/internal/PlatformLive.scala:
```diff
@@ -45,6 +46,8 @@ object PlatformLive {
       def newWeakHashMap[A, B](): JMap[A, B] =
         new HashMap[A, B]()
+
+      val tracing = Tracing(null, TracingConfig.disabled)
```
:+1:

kaishh commented on core/jvm/src/test/scala/scalaz/zio/ZManagedSpec.scala:
```diff
@@ -378,7 +377,7 @@ class ZManagedSpec(implicit ee: org.specs2.concurrent.ExecutionEnv) extends Test
   def timed = {
     val managed = ZManaged(
-      clock.sleep(100.milliseconds) *> ZIO.succeed(Reservation(clock.sleep(100.milliseconds), ZIO.unit))
+      clock.sleep(1000.milliseconds) *> ZIO.succeed(Reservation(clock.sleep(1000.milliseconds), ZIO.unit))
```
Enabled tracing causes these to flap more often... Shouldn't be a problem after the latch fixes.

kaishh commented on core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala:
```diff
@@ -78,13 +162,23 @@ private[zio] final class FiberContext[E, A](platform: Platform, startEnv: AnyRef
+          if (traceStack && inTracingRegion) {
+            popStackTrace(); addStackTrace(k)
+          }
+          stack.push(k)
```
👍

kaishh commented:
@jdegoes 
Ok, I think I've addressed everything, and moved the 2 of the bigger items to the tracker:

- https://github.com/scalaz/scalaz-zio/issues/916 - seemingly no perf gain observed from removing ZIOFn
- https://github.com/scalaz/scalaz-zio/issues/915 - detangling tracing logic while preserving JIT-friendliness might take quite some time

kaishh commented:
:shipit: ?

kaishh commented:
Final benchmarks (on a mid-tier xeon):

enabled tracing:
```
[info] Result "scalaz.zio.IOLeftBindBenchmark.scalazLeftBindBenchmark":
[info]   4985.133 ±(99.9%) 74.398 ops/s [Average]
[info]   (min, avg, max) = (4871.750, 4985.133, 5034.919), stdev = 49.210
[info]   CI (99.9%): [4910.735, 5059.531] (assumes normal distribution)
[info] # Run complete. Total time: 00:03:21
```

disabled tracing:
```
[info] Result "scalaz.zio.IOLeftBindBenchmark.scalazLeftBindBenchmark":
[info]   11538.118 ±(99.9%) 126.977 ops/s [Average]
[info]   (min, avg, max) = (11343.236, 11538.118, 11627.687), stdev = 83.987
[info]   CI (99.9%): [11411.141, 11665.095] (assumes normal distribution)
[info] # Run complete. Total time: 00:03:21
```

cats-effect IO:
```
[info] Result "scalaz.zio.IOLeftBindBenchmark.catsLeftBindBenchmark":
[info]   8210.027 ±(99.9%) 103.346 ops/s [Average]
[info]   (min, avg, max) = (8082.678, 8210.027, 8329.851), stdev = 68.357
[info]   CI (99.9%): [8106.681, 8313.373] (assumes normal distribution)
[info] # Run complete. Total time: 00:03:21
```

Tracing overhead: ~2.3x ; ~1.6x vs cats.

goral09 commented:
@kaishh This looks great. Can't wait to use it. I think it would be great to see example traces. So that people know they're not stacktraces they know and (don't) love :)

kaishh commented:
:tada:

iravid commented:
finally!!

jdegoes commented:
![giphy](https://user-images.githubusercontent.com/156745/58629282-4a51e900-8299-11e9-93a3-33dd9e958343.gif)

kaishh commented:
@goral09 
There are a couple example traces in the presentation at https://www.slideshare.net/jdegoes/error-management-future-vs-zio

Here's also an example trace from fs2 from ZIO test suite:
```
Fiber failed.
A checked error was not handled.
java.lang.Exception
	at scalaz.zio.interop.ZioWithFs2Spec.$anonfun$bracketFail$7(Fs2ZioSpec.scala:61)
	at scalaz.zio.ZIO$ZipRightFn.apply(ZIO.scala:2081)
	at scalaz.zio.ZIO$ZipRightFn.apply(ZIO.scala:2078)
	at scalaz.zio.internal.FiberContext.evaluateNow(FiberContext.scala:685)
	at scalaz.zio.internal.FiberContext.$anonfun$fork$1(FiberContext.scala:573)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

Fiber:20 was supposed to continue to:
  a future continuation at fs2.internal.Algebra$.compileLoop(Algebra.scala:349)
  a future continuation at fs2.internal.Algebra$.compile(Algebra.scala:167)
  a future continuation at scalaz.zio.ZIO.run(ZIO.scala:1026)
  a future continuation at scalaz.zio.interop.CatsEffect.bracketCase(catsjvm.scala:231)
  a future continuation at fs2.Stream$CompileOps.foldChunks(Stream.scala:3921)

Fiber:20 execution trace:
  at fs2.internal.Algebra$.compileLoop(Algebra.scala:276)
  at fs2.internal.CompileScope.interruptibleEval(CompileScope.scala:381)
  at cats.ApplicativeError.attempt(ApplicativeError.scala:68)
  at scalaz.zio.interop.ZioWithFs2Spec.bracketFail(Fs2ZioSpec.scala:61)
  at scalaz.zio.internal.FiberContext$InterruptExit$.apply(FiberContext.scala:151)
  at scalaz.zio.internal.FiberContext$InterruptExit$.apply(FiberContext.scala:142)
  at scalaz.zio.ZIO.ensuring(ZIO.scala:481)
  at scalaz.zio.ZIO.onInterrupt(ZIO.scala:543)
  at scalaz.zio.ZIO.ensuring(ZIO.scala:479)
  at scalaz.zio.internal.FiberContext$InterruptExit$.apply(FiberContext.scala:142)
  at scalaz.zio.ZIO$._IdentityFn(ZIO.scala:1992)
  at scalaz.zio.ZIOFunctions.effectAsyncInterrupt(ZIO.scala:1444)
  at scalaz.zio.ZIOFunctions.effectAsyncInterrupt(ZIO.scala:1440)
  at scalaz.zio.ZIOFunctions.effectAsyncInterrupt(ZIO.scala:1440)
  at fs2.internal.Algebra$.compileLoop(Algebra.scala:229)
  at fs2.internal.Algebra$.compileLoop(Algebra.scala:250)
  at cats.ApplicativeError.attempt(ApplicativeError.scala:68)
  at cats.ApplicativeError.attempt(ApplicativeError.scala:67)
  at fs2.internal.Algebra$.compileLoop(Algebra.scala:229)
  at fs2.internal.Algebra$.compileLoop(Algebra.scala:286)
  at fs2.internal.CompileScope.acquireResource(CompileScope.scala:168)
  at cats.effect.concurrent.Ref$SyncRef.modify(Ref.scala:253)
  at fs2.internal.CompileScope.acquireResource(CompileScope.scala:168)
  at cats.FlatMap.flatten(FlatMap.scala:41)
  at cats.effect.concurrent.Ref$SyncRef.modify(Ref.scala:253)
  at fs2.internal.CompileScope.acquireResource(CompileScope.scala:164)
  at cats.ApplicativeError.attempt(ApplicativeError.scala:68)
  at cats.ApplicativeError.attempt(ApplicativeError.scala:67)
  at scalaz.zio.ZIO.unit(ZIO.scala:905)
  at scalaz.zio.Promise$.done$extension(Promise.scala:124)
  at fs2.internal.Algebra$.compileLoop(Algebra.scala:229)
  at fs2.internal.Algebra$.compileLoop(Algebra.scala:248)
  at fs2.Stream$Compiler$.compile(Stream.scala:3858)
  at fs2.internal.CompileScope$.newRoot(CompileScope.scala:399)
  at fs2.Stream$Compiler$$anon$8.apply(Stream.scala:3863)
  at fs2.Stream$CompileOps.foldChunks(Stream.scala:3921)

Fiber:20 was spawned by:

Fiber:8 was supposed to continue to:
  a future continuation at scalaz.zio.interop.ZioWithFs2Spec.bracketFail(Fs2ZioSpec.scala:67)
  a future continuation at scalaz.zio.ZIO.timeout(ZIO.scala:950)
  a future continuation at scalaz.zio.ZIO.sandbox(ZIO.scala:1053)
  a future continuation at scalaz.zio.ZIO$.scalaz$zio$ZIO$$_succeedRight(ZIO.scala:2071)

Fiber:8 execution trace:
  at scalaz.zio.interop.ZioWithFs2Spec.bracketFail(Fs2ZioSpec.scala:57)
  at scalaz.zio.Promise$.make(Promise.scala:192)
  at scalaz.zio.interop.ZioWithFs2Spec.bracketFail(Fs2ZioSpec.scala:56)
  at scalaz.zio.Promise$.make(Promise.scala:192)
  at scalaz.zio.interop.ZioWithFs2Spec.bracketFail(Fs2ZioSpec.scala:55)
  at scalaz.zio.Promise$.make(Promise.scala:192)

Fiber:8 was spawned by:

Fiber:5 was supposed to continue to:
  a future continuation at scalaz.zio.ZIO.raceWith(ZIO.scala:279)
  a future continuation at scalaz.zio.ZIO.raceWith(ZIO.scala:285)
  a future continuation at scalaz.zio.ZIO.refailWithTrace(ZIO.scala:650)
  a future continuation at scalaz.zio.ZIO.race(ZIO.scala:223)
  a future continuation at scalaz.zio.ZIOFunctions.absolve(ZIO.scala:1462)
  a future continuation at scalaz.zio.ZIOFunctions.unsandbox(ZIO.scala:1470)

Fiber:5 execution trace:
  at scalaz.zio.ZIO.raceWith(ZIO.scala:275)
  at scalaz.zio.Ref$.make(Ref.scala:142)
  at scalaz.zio.ZIO.raceWith(ZIO.scala:274)
  at scalaz.zio.Promise$.make(Promise.scala:192)

Fiber:5 was spawned by: <empty trace>
````
